### PR TITLE
Group GSR Fix 

### PR DESCRIPTION
**Pull request by vcai122**
https://app.shortcut.com/pennlabs/story/3754/gsr-update-v2-book-in-30-minute-intervals

### CODE REVIEW

File: backend/gsr_booking/group_logic.py [hunk 1]
```diff
@@ -42,36 +43,74 @@ def book_room(self, gid, rid, room_name, start, end, user, group):
...
-                    reservation = Reservation.objects.create(
-                        start=start, end=end, creator=user, group=group
-                    )
-                    booking.reservation = reservation
-                    booking.save()
-                    break
-                except APIError:
-                    pass
+            users = self.get_wharton_users(group)
+            credit_id = gsr.lid
```
Comment (judtinzhang) [inline backend/gsr_booking/group_logic.py:51]: Oh wow, this `credit_id` was really poorly design by me a few months ago. Thanks for accommodating for this!

File: backend/gsr_booking/group_logic.py [hunk 2]
```diff
@@ -42,36 +43,74 @@ def book_room(self, gid, rid, room_name, start, end, user, group):
...
-                    )
-                    reservation = Reservation.objects.create(
-                        start=start, end=end, creator=user, group=group
-                    )
-                    booking.reservation = reservation
-                    booking.save()
+            users = self.get_all_users(group)
+            credit_id = CreditType.LIBCAL.value
+
+        total_credits = sum([self.bw.check_credits(usr.user).get(credit_id, 0) for usr in users])
```
Comment (judtinzhang) [inline backend/gsr_booking/group_logic.py:56]: Nit: change `usr` to `user`. Might be a bit easier to read.

File: backend/gsr_booking/group_logic.py [hunk 3]
```diff
@@ -42,36 +43,74 @@ def book_room(self, gid, rid, room_name, start, end, user, group):
...
+        duration = int((end.timestamp() - start.timestamp()) / 60)
+        if total_credits < duration:
+            raise APIError("Not Enough Credits to Book")
+        if duration % 30 != 0:
+            raise APIError("Invalid duration")
```
Comment (judtinzhang) [inline backend/gsr_booking/group_logic.py:61]: Awesome!

File: backend/gsr_booking/group_logic.py [hunk 4]
```diff
@@ -42,36 +43,74 @@ def book_room(self, gid, rid, room_name, start, end, user, group):
...
+                booking.save()
+                start = curr_end
+                duration -= 30
+                if duration <= 0:
                     break
-                except APIError:
-                    pass
+        return reservation
+
+        #         try:
```
Comment (judtinzhang) [inline backend/gsr_booking/group_logic.py:87]: Might be good to delete the code here

File: backend/gsr_booking/group_logic.py [hunk 5]
```diff
@@ -42,36 +43,74 @@ def book_room(self, gid, rid, room_name, start, end, user, group):
...
+        total_credits = sum([self.bw.check_credits(usr.user).get(credit_id, 0) for usr in users])
+        duration = int((end.timestamp() - start.timestamp()) / 60)
+        if total_credits < duration:
+            raise APIError("Not Enough Credits to Book")
+        if duration % 30 != 0:
+            raise APIError("Invalid duration")
+
+        reservation = Reservation.objects.create(start=start, end=end, creator=user, group=group)
+        while duration > 0:
+            for usr in users:
```
Comment (judtinzhang) [inline backend/gsr_booking/group_logic.py:65]: Nit: `user` instead of `usr`

File: backend/gsr_booking/group_logic.py [hunk 6]
```diff
@@ -42,36 +43,74 @@ def book_room(self, gid, rid, room_name, start, end, user, group):
...
+                    start.strftime("%Y-%m-%dT%H:%M:%S%z"),
+                    curr_end.strftime("%Y-%m-%dT%H:%M:%S%z"),
+                    usr.user,
+                    group_book=True,
+                )
+                booking.reservation = reservation
+                booking.save()
+                start = curr_end
+                duration -= 30
+                if duration <= 0:
```
Comment (judtinzhang) [inline backend/gsr_booking/group_logic.py:83]: Good optimization :)

**Review by judtinzhang — CHANGES_REQUESTED** [review]
This looks great! Just a couple of nitpicks. Also, do you mind adding a couple of comments here in the code and tests? Might be good for future developers on the project. Feel free to re-request review after and then we can get this merged :)

File: backend/gsr_booking/group_logic.py (hunk 2 above)
Comment (vcai122) [reply backend/gsr_booking/group_logic.py:56]: i didn't use user since user was an input to the function... would o_user or other be better?

Comment (judtinzhang) [reply backend/gsr_booking/group_logic.py:56]: Oh, I see haha. In that case let's leave it as is :)

**Review by judtinzhang — APPROVED** [review]

**Comment by judtinzhang** [comment]
Good stuff Vincent !

**Pull request merged**